### PR TITLE
doc: graph: document for complex fusions

### PR DESCRIPTION
**Pull request by TaoLv**
1. Created a dedicate document folder for fusion patterns.
2. Add documents for GQA and gated MLP.
3. Improve the existing SDPA document: mention the MQA example, explain the memory consumption for reference implementation, fix wording, etc.
4. Remove legacy graph compiler patterns.

### CODE REVIEW

**Review by vpirogov — APPROVED** [review]

**Comment by vpirogov** [comment]
Looks nice, @TaoLv.

**Review by ElaineBao — APPROVED** [review]

File: doc/graph/fusion_patterns/fusion_patterns.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+B's second input are graph partition input, and they share the same input tensor
+in1. Most input tensors of a graph partition are not explicitly marked.
+For example, the input tensors of the first op are implicitly regarded as graph
+partition inputs. Besides, for input tensors of other ops, if they are not
+produced by any proceeding ops, they are regarded as implicit graph partition
+inputs. In the example A\f$_{>t1}\f$+B+C\f$_{<t1}\f$, A's inputs are
+regarded as implicit graph partition inputs, and if B is a binary operation, the
+second input tensor is an implicit graph partition input.
+
+The following categories will be used in describing fusion pattern.
```
Comment (ranukund) [inline doc/graph/fusion_patterns/fusion_patterns.md:58]:
```suggestion
The following categories will be used in describing a fusion pattern.
```

File: doc/graph/fusion_patterns/gqa.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+oneDNN supports the floating-point GQA pattern with data types f32, bf16, and
+f16. You can specify the data type via the input and output data type fields of
+logical tensors for each operation. oneDNN does not support mixing different
+floating data types in a floating-point GQA pattern.
+
+The definition of the data types and support status on different CPU and GPU
+platforms follow the general description in @ref dev_guide_data_types.
+
+## Implementation limitations
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md:69]:
```suggestion
## Implementation Limitations
```

File: doc/graph/fusion_patterns/gqa.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+The notations used in the document:
+
+- N: the mini-batch size.
+- H_q: the head number of Query.
+- H_kv: the head number of Key or Value.
+- N_rep: H_q / H_kv, indicates how many Query heads are mapped to one Key head.
+- S: the sequence length.
+- D: the size of each head.
+
+## GQA pattern
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md:29]:
```suggestion
## GQA Pattern
```

**Comment by TaoLv** [comment]
> Minor comments on the fusion_patterns.md file. The fusion_patterns.md is an exisiting file which was not reviewed earlier. Please incorporate as you see fit, thanks Tao

Thank you @ranukund . The suggestions are incorporated now.

**Pull request merged**